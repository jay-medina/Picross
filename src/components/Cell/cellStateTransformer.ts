export enum CellState {
  Selected = 'selected',
  Crossed =  'crossed',
}

export interface CellStates {
  [key: string]: (CellState | undefined);
}


export function updateState(oldState: CellStates, row: number, column: number) {
  const key = createKey(row, column);
  const state = getState(oldState, key);

  return {
    ...oldState,
    [key]: toggleState(
      state,
      row,
      column,
    ),
  };
}

export function getState(state: CellStates, key: string) {
  return state[key];
}

export function createKey(row: number, column: number) {
  return `${row} - ${column}`;
}

function toggleState(state: CellState | undefined, row: number, column: number) {
  switch (state) {
    case CellState.Selected: return CellState.Crossed;
    case CellState.Crossed: return undefined;
    default: return CellState.Selected;
  }
}
