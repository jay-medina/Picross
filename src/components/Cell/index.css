.cell {
  width: 2.5rem;
  height: 2.5rem;
  border: 1px solid;
  border-color: var(--dark-blue);
  font-size: 1rem;
  display: flex;
  justify-content: center;
  background-color: var(--board-color);
  cursor: pointer;
}

.cell .selected {
  background-color: var(--white);
  margin: 0.15rem;
  border-radius: 3px;
  width: inherit;
}

.cell .cross {
  color: var(--secondary-color);
  font-size: 2rem;
}

.cell.hint {
  font-size: 1.35rem;
  padding-top: 0.25rem;
  color:white;
  background-color: var(--secondary-color);
}

.cell.border-cell-left {
  border-right-color: white;
}

.cell.border-cell-right {
  border-left-color: white;
}
