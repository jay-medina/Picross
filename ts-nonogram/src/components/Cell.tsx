import React from 'react';

const Cell: React.StatelessComponent<{}> = () => {
  return <div>Hello world</div>;
};

export default Cell;
