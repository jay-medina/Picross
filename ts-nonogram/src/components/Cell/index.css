:root {
  --main-color: #546e7a;
  --secondary-color: #29434e;
  --white: #FFF;
}
.cell {
  width: 2.5rem;
  height: 2.5rem;
  border: 1px solid;
  font-size: 1rem;
  display: flex;
  justify-content: center;
  background-color: var(--main-color);
}

.selected {
  background-color: var(--white);
  margin: 0.15rem;
  width: inherit;
}

.cross {
  color: var(--secondary-color);
  font-size: 4rem;
  position: relative;
  top: -17px;
}
