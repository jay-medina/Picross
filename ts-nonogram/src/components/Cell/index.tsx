import React from 'react';
import './index.css';

const Cell: React.StatelessComponent<{}> = () => {
  return <div className="cell">Hello world</div>;
};

export default Cell;
