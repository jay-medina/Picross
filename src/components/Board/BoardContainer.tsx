import React from 'react';
import { CellState } from '../Cell';
import Board, { CellStates } from './index';

export interface BoardContainerProps {
  rows: number;
  columns: number;
}


export interface BoardContainerState {
  cellStates: CellStates;
}

class BoardContainer extends React.PureComponent<BoardContainerProps, BoardContainerState> {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
    this.state = {
      cellStates: {},
    };
  }

  render() {
    return <Board 
      {...this.props} 
      onClick={this.onClick}
      cellStates={this.state.cellStates}
    />;
  }

  private onClick(row: number, column: number) {
    const key = `${row} - ${column}`;

    const newCellStates = {
      ...this.state.cellStates,
      [key]: this.toggleStates(
        this.state.cellStates[key],
      ),
    };

    this.setState({
      cellStates: newCellStates,
    });
  }

  private toggleStates(state?: CellState) {
    switch (state) {
      case CellState.Selected: return CellState.Crossed;
      case CellState.Crossed: return undefined;
      default: return CellState.Selected;
    }
  }

}

export default BoardContainer;
