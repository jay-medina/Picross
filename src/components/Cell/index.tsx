import React from 'react';
import './index.css';
import { CellState } from './cellStateTransformer';

export interface CellProps {
  state?: CellState;
  row: number;
  column: number;
  totalRows: number;
  totalColumns: number;
  onClick: (row: number, column: number) => void;
}

class Cell extends React.PureComponent<CellProps, {}> {
  constructor(props: CellProps) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }
  render() {
    return (
      <div className={this.getClassName()} onClick={this.onClick}>
        {this.getChildElement()}
      </div>
    );
  }

  private getClassName() {
    return `cell ${this.borderColumnCell()} ${this.borderRowCell()}`;
  }

  private borderColumnCell() {
    const { column, totalColumns } = this.props;
    if (column < 4 || totalColumns < 6) 
      return '';

    if ((column + 1) % 5 === 0) {
      return 'border-cell-right';
    } 
    
    if (column % 5 === 0) {
      return 'border-cell-left';
    }
    return '';
  }

  private borderRowCell() {
    const { row, totalRows } = this.props;
    if (row < 4 || totalRows < 6) 
      return '';

    if ((row + 1) % 5 === 0) {
      return 'border-cell-bottom';
    } 
    
    if (row % 5 === 0) {
      return 'border-cell-top';
    }
    return '';
  }

  private getChildElement() {
    const { state } = this.props;
    switch (state) {
      case 'crossed': return <span className="cross">&times;</span>;
      case 'selected': return <div className="selected" />;
      default: return null;
    }
  }
  private onClick() {
    this.props.onClick(this.props.row, this.props.column);
  }
}

export default Cell;
