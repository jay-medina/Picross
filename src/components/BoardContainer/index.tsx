import React from 'react';
import Board from '../Board';
import { CellStates, updateState } from '../Cell/cellStateTransformer';

export interface BoardContainerProps {
  rows: number;
  columns: number;
  rowHints: number[][];
  columnHints: number[][];
}


export interface BoardContainerState {
  cellStates: CellStates;
}

class BoardContainer extends React.PureComponent<BoardContainerProps, BoardContainerState> {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
    this.state = {
      cellStates: {},
    };
  }

  render() {
    return <Board
      {...this.props}
      onClick={this.onClick}
      cellStates={this.state.cellStates}
    />;
  }

  private onClick(row: number, column: number) {
    const newCellStates = updateState(
      this.state.cellStates,
      row,
      column,
    );

    this.setState({
      cellStates: newCellStates,
    });
  }

}

export default BoardContainer;
