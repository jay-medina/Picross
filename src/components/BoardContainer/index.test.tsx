jest.mock('../Cell/cellStateTransformer');

import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import BoardContainer from '.';
import { updateState, CellState } from '../Cell/cellStateTransformer';

describe('<BoardContainer />', () => {
  let wrapper: ShallowWrapper;

  describe('when the container is created', () => {
    beforeEach(() => {
      wrapper = shallow(
        <BoardContainer rows={2} columns={2} rowHints={[]} columnHints={[]} />,
      );
    });

    it('renders the board with rows and columns', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when the board is clicked', () => {
    beforeEach(() => {
      wrapper = shallow(
        <BoardContainer rows={2} columns={2} rowHints={[]} columnHints={[]} />,
      );
      (updateState as jest.Mock).mockReturnValue({
        '0 - 0': CellState.Selected,
      });
      wrapper.find('Board').simulate('click', 0, 0);
    });

    it('updates the cell state', () => {
      expect(updateState).toHaveBeenCalledWith({}, 0, 0);
    });

    it('sets the new cell state on the board container', () => {
      expect(wrapper.find('Board').prop('cellStates')).toEqual({
        '0 - 0': CellState.Selected,
      });
    });
  });
});
