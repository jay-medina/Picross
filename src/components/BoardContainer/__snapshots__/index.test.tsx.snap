// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardContainer /> when the container is created renders the board with rows and columns 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BoardContainer
    columnHints={Array []}
    columns={2}
    rowHints={Array []}
    rows={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "cellStates": Object {},
      "columnHints": Array [],
      "columns": 2,
      "onClick": [Function],
      "rowHints": Array [],
      "rows": 2,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "cellStates": Object {},
        "columnHints": Array [],
        "columns": 2,
        "onClick": [Function],
        "rowHints": Array [],
        "rows": 2,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
